torch/distributed/fsdp/_utils.py

```diff
@@ -78,3 +83,51 @@ def f(module: torch.nn.Module, prefix: str, *args, **kwargs):
 
     f(root_module, "", *args, **kwargs)
     return return_fn(*args, **kwargs)
+
+
+@torch.no_grad()
+def _alloc_storage(tensor: torch.Tensor, size: torch.Size) -> bool:
+    """
+    Allocate storage for ``tensor`` with the given size.
+
+    Returns:
+        bool: ``True`` if this method allocated storage and ``False`` if the
+        storage was already allocated.
+    """
+    already_allocated = tensor.storage().size() == size.numel()
+    if not already_allocated:
+        tensor_storage_size = tensor.storage().size()
+        p_assert(
+            tensor_storage_size == 0,
+            f"Tensor storage should have been resized to be 0 but got {tensor_storage_size}",
+        )
+        tensor.storage().resize_(size.numel())
+    return not already_allocated
+
+
+@torch.no_grad()
+def _free_storage(tensor: torch.Tensor) -> bool:
+    """
+    Frees the underlying storage of ``tensor``.
+
+    Returns:
+        bool: ``True`` if the method freed the storage and ``False`` if the
+        storage was already freed.
+    """
+    already_freed = tensor.storage().size() == 0
+    if not already_freed:
+        p_assert(
+            tensor.storage_offset() == 0,
+            "Freeing a tensor's storage is unsafe when it is not the sole occupant",
+        )
+        tensor.storage().resize_(0)
```

Comment:
Will this allow the underlying memory block to be fully reused? Or does the allocator have to split the block to preserve its index block ptr? Or is there a special handling underneath which treats resize_(0) as a deallocation?

Reply:
Given the logic I outlined in the other comment, I think `resize_(0)` is simply a deallocation: https://github.com/pytorch/pytorch/pull/83665#discussion_r955588825

I am not sure what you mean by the "index block ptr", so let me know if the linked comment helps answer your question or if it is still unanswered.